### ISSUE TITLE
How to pass parameters to an ActiveX?

### ISSUE THREAD

**Issue by rejurime** [dolphinsmalltalk/Dolphin#982]
Hi!
I am having a problem sending a list of points to an ActiveX component that I have as an example.
The component defines the following code:

MFCActiveXControl1.idl file
```cpp
dispinterface _DMFCActiveXControl1
{
  properties:
  methods:
    [id (161)] void PrintPolygonDoubles (VARIANT * xPoints, VARIANT * yPoints, long numPoints);
```

MFCActiveXControl1Ctrl.h file
```cpp
afx_msg void PrintPolygonDoubles (VARIANT * xPoints, VARIANT * yPoints, long numPoints);
```

MFCActiveXControl1Ctrl.cpp file
```cpp
BEGIN_DISPATCH_MAP (CMFCActiveXControl1Ctrl, COleControl)
  DISP_FUNCTION_ID (CMFCActiveXControl1Ctrl, "PrintPoligonoDoubles", 161, PrintPoligonoDoubles, VT_EMPTY, VTS_VARIANT VTS_VARIANT VTS_I4)
END_DISPATCH_MAP ()

void CMFCActiveXControl1Ctrl :: PrintPolygonDoubles (VARIANT * xPoints, VARIANT * yPoints, long numPoints)
{
  SAFEARRAY sax = xPoints-> pparray;
  SAFEARRAY * say = * yPoints-> pparray;
  double * xPts = (double *) sax-> pvData;
  double * yPts = (double *) say-> pvData;
  char x = 'X';
  char y = 'Y';

  if (numPoints)
  {
    ofstream fs ("points.txt");
    register int i;
    for (i = 0; i <numPoints; i ++)
    {
      fs << x << endl;
      fs << xPts [i] << endl;
      fs << and << endl;
      fs << yPts [i] << endl;
    }
    fs.close ();
  }
}
```

Installing the OCX component with the Dolphin wizard works without problems and the MFCActiveXControl1Lib_DMFCActiveXControl1 class is created that contains the printPoligonoDoubles: xPoints andPoints: yPoints numPoints: numPoints method. I create a new instance of the class in the AXControlSite container and then when I execute the method the following happens to me:
```smalltalk
aView controlDispatch printPolygonDoubles: # (12.5 12.7) andPoints: # (16.5 21.7) numPoints: 2 -> Returns "Invalid access to memory location"
```
If I change the run configuration to:
```smalltalk
x: = VARIANT new arrayRef: # (12.5 12.7).
y: = VARIANT new arrayRef: # (16.5 21.7).
aView controlDispatch printPolygonDoubles: x yPoints: y numPoints: 2
```
Run perfectly and add the following to the file:
```
X
2.47033e-323
AND
2.47033e-323
X
12.5
AND
16.5
```
that is, it adds a zero in front of each number.
If we execute the same but passing a 4 instead of 2 as numPoints
```smalltalk
x: = VARIANT new arrayRef: # (12.5 12.7).
y: = VARIANT new arrayRef: # (16.5 21.7).
aView controlDispatch printPolygonDoubles: x yPoints: y numPoints: 4
```
Run perfectly and add the following to the file:
```
X
2.47033e-323
AND
2.47033e-323
X
12.5
AND
16.5
X
2.47033e-323
AND
2.47033e-323
X
12.7
AND
21.7
```
The specific question is how can I pass the list of numbers to the library correctly? or I'm doing it right and it's a Dolphin problem.
The version of Dolphin I am using is: 7.1.9.

Any suggestion?

Thank you very much for your time.
Regards.

**Comment by blairmcg** [comment]
This is more of a support request than a Dolphin issue, so I'll be closing it with my comment. If you encounter specific issues that are or might be problems in Dolphin then by all means raise them, but otherwise there is nothing specific here that suggests it is appropriate to raise an issue.

The control has a very loosely defined interface, but then is expecting VARIANTs populated in a specific way. There is no way that Dolphin or the component wizard could deduce this because that information is only in the code, not in the type library or dynamic type information. If you ask Dolphin to translate an arbitrary Array to a COM SAFEARRAY, it will build a SAFEARRAY of VARIANTs, because that is the most general form. Judging from the control source code clip you have attached, the control is expecting (without checking) that it is passed array references in the wrapping VARIANT, and that the array referenced is a SAFEARRAY of doubles, not VARIANTs. To create SAFEARRAY of doubles in Dolphin you will have to specifically request it, e.g:
```
x := VARIANT new arrayRef: (SAFEARRAY withAll: #(12.5 12.7) elementClass: DOUBLE).
y: = VARIANT new arrayRef: (SAFEARRAY withAll: #(16.5 21.7) elementClass: DOUBLE).
aView controlDispatch printPolygonDoubles: x yPoints: y numPoints: 2.
```
The control may work with VB(A) because it happens to use particular underlying representations anyway, but really it should either have a much more strongly typed interface, or it should be checking the types of the parameters it actually receives and be prepared to validate those and/or convert them to what it wants. Based on that, I would be wary of its quality.

I suspect you are going to run into other problems with this control that will require you to debug it to determine similar assumptions it is making. Since you have the control source code, you should build it in Debug with VisualStudio, Attach the VS debugger to Dolphin, and then set a breakpoint in the control's function and see what you are being passed in. Then you can experiment and try and understand what is being passed in when you do different things on the Dolphin side. It will be difficult to make rapid progress if you stick to examining the end result by looking at what ends up being written to a file - that's akin to printf debugging really.

**Comment by rejurime** [comment]
My bad, thank you very much for taking the time to reply. I ended up creating an issue as a question because I saw that the [Gitter](https://gitter.im/objectarts/dolphinsmalltalk) that I used in the past, didn't have activity. But I'll keep it in mind for the future, I assume that for questions of this type, the right channel to ask is the google group (comp.lang.smalltalk.dolphin) right? Once again, thank you for your time and work.
